### ISSUE TITLE
Kata Editor Page Blanking Out When I Press Command-V

### ISSUE THREAD

**Issue by dpleshkov** [Codewars/codewars.com#828]
As the title says, the kata editor page is blanking out when I try to paste a code snippet in the description field of the kata editor.

The code snippet:
```
div {
    display: none;
}
    
a:hover + div {
    display: block;
}
```
And where I try to place it:
```
Some description text here
<style>
Where I paste it.
</style>
```

I think the problem is that I am trying to hide a certain part of the kata's description to hide a hint but the browser page thinks that I am changing its HTML. I am doing this with Mac OS Sierra, on Chrome.

**Comment by kazk** [comment]
I'm not sure, but isn't this because of the selector `div`? The rule is applied to **every** `<div>` on the page.

I'd recommend using inline styles for kata descriptions and avoid "type selectors" in general.
Unfortunately, `:hover` wouldn't work inline. How about using the same color for `color` and `background-color`? Haven't tested, but this should *hide* the hint unless selected.

```html
Some description text here

Some hint below:
<p style="color: black; background-color: black">hint</p>
```

If you really need `:hover`, using class selector (`.kata-hint`) might work.

---

@jhoffner, `<style></style>` in the description seems to affect the entire page. I saw this recently on [6 By 6 Skyscrapers](https://www.codewars.com/kata/6-by-6-skyscrapers)'s page.

![image](https://cloud.githubusercontent.com/assets/639336/22942988/9781ffe8-f2a0-11e6-93d0-f9688ad07fbb.png)

I think this is caused by following `<style>` in the description.

```html
<style type="text/css">
table { width: 276px; }
table, tr, td { border: 0px; }
</style>
```

I like how Codewars allows authors to do this because the description of that kata looks nice, but maybe document this behavior or add some warning?

**Comment by kazk** [comment]
@jhoffner, this is caused by `div { display: none; }` applied to `<div id="app"></div>`. 

Pasting `div { display: table or flex or inline; }` will produce similar observable change, but `display: none` is more serious issue because it doesn't allow to recover.

```css
div { display: none }
/*                 ^ 'backspace' after typing that last 'e' won't make the editor go back */
```

The only way I found to recover is to remove that rule through Developer Tools.

**Comment by jhoffner** [comment]
Yeah `div { display: none}` is going to make every div element on the page invisible. The markdown isn't somehow scoped or sandboxed. I'm not really sure what the fix here is to prevent people from doing this, all I can think of is to just forbid style tags all together. 

For what its worth. `div {display: none;}` is not likely to ever be something you would want to do within any programming context, unless it is at least scoped (ie: `.hide-divs div {display: none;}`)

Is it enough to just say lets no do this anymore?

**Comment by jhoffner** [comment]
I suppose I could try to parse out the style tag and add `.markdown ` as a prefix to any css line, but that seems like likely to have all kinds of edge cases.

**Comment by kazk** [comment]
> The markdown isn't somehow scoped or sandboxed.

What about JavaScript?

> I'm not really sure what the fix here is to prevent people from doing this, all I can think of is to just forbid style tags all together.

I'm against forbidding because it does make the description **much** easier to read if used properly and many descriptions will break.
I don't know how the kata editor is implemented, but if it's triggering an event whenever user edits, then maybe lightly check the content in the event handler?
I'll try to look into this and let you know if I find anything.

**Comment by dpleshkov** [comment]
Maybe I should just make the selector `div.hint` and classify the div tag as `hint`.

**Comment by dpleshkov** [comment]
I also think you should still be able to style the description page.

**Comment by kazk** [comment]
@jhoffner 

I've done some experiments and read some of `application.js`. I don't think it's difficult to modify `App.Controls.MarkdownEditor` (`refresh`) and `App.Controls.MarkdownDisplay` (`setMarkdown`) to do whatever you decide. I need to read more `marked` documents, but it has some useful options and its lexer and parser are available to use if that doesn't suffice.

---

I believe I've found more serious issue which I don't know if I should post here. I have no experience in security and it might not be a risk at all, but it doesn't look safe to me.
Should I email you?

**Comment by jhoffner** [comment]
@kazk you can send me the details on the security risk at jake@codewars.com.

**Comment by kazk** [comment]
@jhoffner I actually sent an email to info@codewars.com 2 days ago. I'll try sending it to your address too.
